ReSharper.FSharp/src/FSharp/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs

```diff
@@ -19,78 +22,100 @@ open JetBrains.ReSharper.Plugins.FSharp.TypeProviders.Protocol
-
-    let connect requestingProjectOutputPath =
-        if isConnectionAlive () then true else
+    let analyzeGenerativeTypeProvidersInMemory = lazy experimentalFeatures.GenerativeTypeProvidersInMemoryAnalysis.Value
+    let clients = ConcurrentDictionary<TypeProvidersHostingScope, ITypeProvidersClient Lazy>()
 
-        lock createProcessLockObj (fun () ->
-            if isConnectionAlive () then true else
+    let getSolutionTypeProvidersClient () =
+        let client = clients.GetValueSafe(TypeProvidersHostingScope.Solution)
+        if isNull client then null.As<ITypeProvidersClient>() else client.Value
 
-            typeProvidersHostLifetime <- Lifetime.Define(lifetime)
-            let isInternalMode = productConfigurations.IsInternalMode()
-            let externalProcess =
-                typeProvidersProcessFactory.Create(
-                        typeProvidersHostLifetime.Lifetime,
-                        requestingProjectOutputPath,
-                        isInternalMode)
+    let terminateConnections () =
+        for client in clients.Values do
+            client.Value.Terminate()
 
-            if isNull externalProcess then false else
-            let newConnection = externalProcess.Run()
+        clients.Clear()
 
-            typeProvidersManager <- TypeProvidersManager(newConnection, fcsProjectProvider, scriptPsiModulesProvider,
-                                                         outputAssemblies, generativeTypeProvidersInMemoryAnalysisEnabled.Value) :?> _
-            connection <- newConnection
-            true)
+    let rec connect (scope: TypeProvidersHostingScope) (resolutionEnv: ResolutionEnvironment) =
+        match clients.TryGetValue(scope) with
+        | true, client ->
+            if client.Value.IsActive then
+                client.Value
+            else
+                clients.TryRemove(scope) |> ignore
+                connect scope resolutionEnv
+        | _ ->
+
+        let client: ITypeProvidersClient Lazy =
+            lazy
+                let clientLifetimeDef = Lifetime.Define(lifetime)
+                let isInternalMode = productConfigurations.IsInternalMode()
+                let logPrefix = scope.ToString()
+
+                let externalProcess =
+                    typeProvidersProcessFactory.Create(
+                        clientLifetimeDef.Lifetime,
+                        Option.toObj resolutionEnv.OutputFile,
+                        isInternalMode, logPrefix)
+
+                let connection = externalProcess.Run()
+
+                match scope with
+                | Solution ->
+                    SolutionTypeProvidersClient(clientLifetimeDef, connection, fcsProjectProvider, outputAssemblies,
+                                                analyzeGenerativeTypeProvidersInMemory.Value) :> _
+                | Scripts ->
+                    ScriptTypeProvidersClient(clientLifetimeDef, connection, scriptPsiModulesProvider) :> _
```

Comment:
What if two projects in a solution use conflicting packages?

Reply:
https://github.com/dotnet/fsharp/issues/11135